src/messaging/messaging.ts

```diff
@@ -228,6 +235,55 @@ export class Messaging implements MessagingInterface {
     return this.appInternal;
   }
 
+  /**
+   * Converts from the public {@link messaging.Message `Message`} type to the
+   * internal/generated {@link protos.google.firebase.fcm.v1.Message `clientMessage`}
+   * type.
+   *
+   * @param message The {@link messaging.Message `Message`} to be converted
+   * @returns A {@link protos.google.firebase.fcm.v1.Message `clientMessage`},
+   * where like fields are the same as the inptuted message.
+   */
+  private convertToClientMessage(message: Message): ClientMessage {
+    //TODO: Add conversion for android, webpush, apns
+
+    const clientMessage: ClientMessage = {
+      data: message.data,
+      notification: message.notification,
+      fcmOptions: message.fcmOptions
+    }
+
+    if ('token' in message) {
+      clientMessage.token = message.token;
+    } else if ('topic' in message) {
+      clientMessage.topic = message.topic;
+    } else if ('condition' in message) {
+      clientMessage.condition = message.condition;
+    }
+
+    return clientMessage;
+  }
+
+  /**
+   * Method that gets the current App's credentials
+   * @returns The credential key of the current Firebase App
+   */
+  private getAppKey(): object{
```

Comment:
This should be able to handle application default credentials too. I'd recommend implementing this as follows:

```ts
private getClientOptions(): object {
  const options = {fallback: 'rest'};
  // If service account credentials, add "credentials" to options.
  return options;
}
```